### ISSUE TITLE
Recent changes to mock.pyi incompatible with legimate uses of patch decorator

### ISSUE THREAD

**Issue by erictraut** [python/typeshed#4566]
The recent changes to mock.pyi have caused spurious errors for legitimate code.

The code for the `@patch` decorator no longer supports idioms like the following:
```python
from mock import patch
import math

@patch("math.pi", 3.14)
def func():
    print(math.pi)

func()
```

The implementation seems to assume that the second argument, if provided, is always a callable, so the above code causes the type checker to emit errors.

Also, it doesn't support the following common case:
```python
@patch("internal_module.send_message", Mock())
def test_func(self):
     do_something_to_trigger_send_message()
```

The above code is flagged as an error because the `__call__` method in `Mock` has a var-arg parameter, and the decorated function does not, so the latter cannot be assigned to the former.

In short, the changes to mock.pyi have broken many cases where an argument is passed for the `new` parameter to `patch`. 

@srittau, perhaps these changes should be reverted until a fix is identified?

**Comment by srittau** [comment]
Oh, it seems I mistyped `_patch.__call__()` (by reusing the `_T` type var). I will submit a PR.

**Comment by erictraut** [comment]
@srittau Thanks for the quick fix! I confirmed that it works.
